### DISCUSSION TITLE
setting local time on pico w

### DISCUSSION THREAD

**Discussion by diegorebollo** [micropython/micropython#9545] [Hardware & Peripherals] [answered]
As far i know the way to set the time on a pico w is using ntptime module, this works fine but i want to set my time zone and i don´t know how to do it. Is there any "official" way for that?

**Comment by jimmo** [comment, marked as answer]
I think it makes sense for ntptime to set the time in UTC. The issue is that micropython itself has no built-in notion of timezones or the current timezone, so `time.localtime` and `time.gmtime` will always return the same thing.

In micropython-lib we have a (micro) datetime package. I have not looked into it in detail, but I believe that as long as your app code knows the current timezone, then you can construct offset times. e.g. assuming the clock has been set by ntptime

```py
my_timezone = timezone(timedelta(hours=10))
current_time = datetime.datetime.now(my_timezone)
```

https://github.com/micropython/micropython-lib/tree/master/python-stdlib/datetime

(You can install this via mip if you're using the latest `mpremote` (0.4.0) or nightly build of the pico w firmware)

**Reply by diegorebollo** [reply]
Thanks! I just tried using datetime as you suggests and everything works as expected.

**Reply by jimmo** [reply]
That's great to know! Thanks

**Reply by profhuster** [reply]
Here is a module I wrote. It defaults to Eastern Standard Time, UTC-5. It is much simpler than one of the other posted solutions.
```
"""
getNow.py
ProfHuster@GMail.com
2022-12-12

Usage (assuming a network connection):
>>> from getNow import getNow
>>> print(getNow(UTC_OFFSET=-5))
(2022, 12, 12, 10, 34, 34, 0, 346)

"""

import time
import ntptime
try:
    ntptime.settime()
except:
    raise Exception("npttime.settime() failed. No network connection.")

def getNow(UTC_OFFSET=-5):
    return time.gmtime(time.time() + UTC_OFFSET * 3600)

if __name__ == "__main__":
    print(getNow())
```

**Comment by GM-Script-Writer-62850** [comment]
This was my solution:
`from TIME import time`
TIME.py:
```
#!/usr/bin/python3
import time as TIME
class time:
	tz=-4
	tz_sec=tz*3600
	tz_ms=tz_sec*1000
	tz_ns=tz_ms*1000000
	gmtime=TIME.gmtime
	def localtime(t=0):
		if not t:
			t=time.time()
		t+=time.tz_sec
		return TIME.localtime(t)
	mktime=TIME.mktime
	sleep=TIME.sleep
	sleep_ms=TIME.sleep_ms
	ticks_us=TIME.ticks_us
	ticks_cpu=TIME.ticks_cpu
	ticks_add=TIME.ticks_add
	ticks_diff=TIME.ticks_diff
	time=TIME.time
	#def time():
	#	return TIME.time()+time.tz_sec
	def time_ms():
		return round(time.time_ns()/1000000)
	time_ns=TIME.time_ns
	#def time_ns():
	#	return TIME.time_ns()+time.tz_ns
	def tzset(tz):
		time.tz=int(tz)
		time.tz_sec=tz*3600
		time.tz_ms=tz_sec*1000
		time.tz_ns=tz_ms*1000000
```
-4 is the default timezone, easily change in source or set on the fly via `time.tzset(8)` where 8 is the hours to add/subtract
time.localtime() will adjust the time given for the local time if you need a converted time stamp `time.time()+time.tz_sec`

**Reply by clach04** [reply]
This only works for whole hour timezones (for example won't work for India). I like the monkey-patch-like approach 👍

**Comment by peterhinch** [comment]
See [this version of ntptime](https://github.com/peterhinch/micropython-samples/blob/master/README.md#414-ntp-time).

**Comment by beetlegigg** [comment]
Yet another option is to get the time from http://worldtimeapi.org which can return the time in json format for any location including the daylight saving time if its in operation.

**Comment by Zorbadil** [comment]
I don't know if this helps at all, but I simply took the tuple from rtc.datetime() and converted it to a list called tlist, changed the hour element in the list and initalised the rtc  with the list using rtc.init(tlist).  This works fine on the esp32.  As the 2040 has an RTC I would assume it would work on it too.

RTC imported as:
from machine import RTC
rtc = RTC()
assuming wirelss connection achieved:

ntptime.settime()
tlist = list(rtc.datetime())
tlist[4] = tlist[4] + 1 # BST in the UK is UTC + 1
rtc.init(tlist)

then use rtc.datetime() as the time source.

**Reply by GM-Script-Writer-62850** [reply]
You could use `tlist[4] +=  1` to make it shorter

**Reply by Zorbadil** [reply]
I wasn't aware of that, being fairly new to python and micropython.  I'd tried the C syntax Variable++ for other stuff and found it didn't work, so I lazily drew the line at that and went back to what I was almost certain would work.  My coding at the moment is fuctional, but not very elegant.

Thanks for the tip though.  It's much nicer and I'm sure my OCD will appreciate it for future use.

**Reply by GM-Script-Writer-62850** [reply]
i'm sure i have done the same thing not being sure if it would work in a given language

sadly `tlist[4]++` does not work in python

**Reply by peterhinch** [reply]
@Zorbadil Does this solution work when the UTC time is 23:xx?

**Reply by vodkawasserfall** [reply]
tlist[4] = ( tlist[4] + 1 ) % 24
@Zorbadil @peterhinch

**Reply by peterhinch** [reply]
Well, yes. Rollover into date, weekday etc.?

**Comment by Zorbadil** [comment]
No.  It's in it's a very crude form of an immediate solution to a problem.  Taking the time accross the date line in the change, and changing/not changine the time based on the date are a couple of the things yet to be addressed.

**Reply by peterhinch** [reply]
See [my take on this](https://github.com/peterhinch/micropython-samples/blob/master/ntptime/ntptime.py).

**Reply by clach04** [reply]
Opened:
  * https://github.com/peterhinch/micropython-samples/issues/49

**Reply by peterhinch** [reply]
As stated in the docs, code in my `micropython-samples` repo is intended for others to adapt and build on. I don't generally respond to feature requests. If others wish to publish a fully supported, more full-featured implementation, I will happily link to it.

**Comment by projectgus** [comment]
I've started a Discussion here to talk about the overall problem and possible ways forward for it:
https://github.com/orgs/micropython/discussions/12378

**Comment by viji2ks** [comment]
'''
file name :  getNow.py
altered for local time with float like in india is utc = +5.5 hours ahead of globa time 
'''

import time
import ntptime
try:
    ntptime.settime()
except:
    raise Exception("npttime.settime() failed. No network connection.")

def getNow(UTC_OFFSET=+5.5):
    local_time = int(UTC_OFFSET * 3600)
    return time.gmtime(time.time() + local_time)

if __name__ == "__main__":
    print(getNow())
    
    
"""
how to use ?

copy whole content  to file and save as getNow.py
viji2ks@gmail.com for any contact
16-09-2023

Usage (assuming a network connection):
>>> from getNow import getNow
>>> print(getNow(UTC_OFFSET=+5.5))
(2022, 12, 12, 10, 34, 34, 0, 346)
(  0    1   2   3   4   5  6   7 )
year, month, day, wday, hour, minute, second, ms_ = rtc.datetime()
# read current datetime from RTC

Index	Attribute	Values
0	tm_year	(for example, 1993)
1	tm_mon	range [1, 12]
2	tm_mday	range [1, 31]
3	tm_hour	range [0, 23]
4	tm_min	range [0, 59]
5	tm_sec	range [0, 61]
6	tm_wday	range [0, 6], Monday is 0
7	tm_yday	range [1, 366]
8	tm_isdst	0, 1 or -1

"""

**Comment by firehurtnova11** [comment]
Thank you for this solution[.](https://vancedmanager.org/)
